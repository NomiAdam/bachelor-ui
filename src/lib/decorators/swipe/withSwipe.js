import React, {PureComponent} from 'react';
import styled from 'styled-components';

const StyledWrapper = styled.div`
  height: 100%;
  width: 100%;
  z-index: 999;
`;

const withSwipe = (Component) => class extends PureComponent {
    state = {
        left: 0,
        originalOffset: 0,
        velocity: 0,
        timeOfLastDragEvent: 0,
        touchStartX: 0,
        prevTouchX: 0,
        beingTouched: false,
        touchStartY: 0,
        isOpen: false,
    };

    componentDidUpdate(prevProps) {
        const {swipeOpen} = this.props;
        if (swipeOpen !== prevProps.swipeOpen) {
            this.setState({
                left: swipeOpen ? 100 : 0,
                isOpen: swipeOpen,
            })
        }
    };

    handleStateChange = () => {
        const {isOpen} = this.state;
        const {handleSwipeChange} = this.props;
        handleSwipeChange(isOpen);
    };

    handleStart = (clientX, clientY) => {
        this.setState({
            originalOffset: this.state.left,
            velocity: 0,
            timeOfLastDragEvent: Date.now(),
            touchStartX: clientX,
            touchStartY: clientY,
            beingTouched: true,
            intervalId: null
        });
    };

    isSwipeX = (clientY) => {
        const {touchStartY} = this.state;
        return Math.abs(touchStartY - clientY) < 5;
    };

    handleMove = (clientX) => {
        if (this.state.beingTouched) {
            const touchX = clientX;
            const currTime = Date.now();
            const elapsed = currTime - this.state.timeOfLastDragEvent;
            const velocity = 20 * (touchX - this.state.prevTouchX) / elapsed;
            let deltaX = touchX - this.state.touchStartX + this.state.originalOffset;
            if (deltaX < 0) {
                deltaX = 0;
            } else if (deltaX > 100) {
                deltaX = 100;
            }
            this.setState({
                left: deltaX,
                velocity,
                timeOfLastDragEvent: currTime,
                prevTouchX: touchX
            });
        }
    };

    handleEnd = () => {
        const {left} = this.state;
        let finalLeft = left;
        if (finalLeft < 50) {
            finalLeft = 0;
        } else if (finalLeft >= 50) {
            finalLeft = 100;
        }
        this.setState({
            velocity: this.state.velocity,
            touchStartX: 0,
            beingTouched: false,
            left: finalLeft,
            isOpen: finalLeft === 100,
        }, this.handleStateChange);
    };

    handleTouchStart = (touchStartEvent) => {
        this.handleStart(
            touchStartEvent.targetTouches[0].clientX,
            touchStartEvent.targetTouches[0].clientY,
        );
    };

    handleTouchMove = (touchMoveEvent) => {
        if (this.isSwipeX(touchMoveEvent.targetTouches[0].clientY)) {
            this.handleMove(touchMoveEvent.targetTouches[0].clientX);
        }
    };

    handleTouchEnd = () => {
        this.handleEnd();
    };

    handleClick = (left) => this.setState({
        left,
        isOpen: left === 100,
    }, this.handleStateChange);

    render() {
        const {left} = this.state;
        const {handleSwipeChange, swipeOpen, ...rest} = this.props;
        return (
            <StyledWrapper
                onTouchStart={this.handleTouchStart}
                onTouchMove={this.handleTouchMove}
                onTouchEnd={this.handleTouchEnd}
            >
                <Component left={left} handleClick={this.handleClick} {...rest} />
            </StyledWrapper>
        );
    }
};

export default withSwipe;
