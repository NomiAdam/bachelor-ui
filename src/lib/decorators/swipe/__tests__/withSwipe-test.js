/* eslint-disable no-undef */
import React from 'react';
import withSwipe from '../withSwipe';
import { shallow } from 'enzyme';

const mockComponent = <div/>;

const Swipe = withSwipe(mockComponent);

describe('withSwipe', () => {
    it('should be defined', () => {
        expect(withSwipe).toBeDefined();
    });
    it('should be defined with component', () => {
        expect(Swipe).toBeDefined();
    });
    it('should match the snapshot', () => {
        expect(shallow(<Swipe />)).toMatchSnapshot();
    });
    it('should have default state defined', () => {
        expect(shallow(<Swipe />).state()).toEqual({
            left: 0,
            originalOffset: 0,
            velocity: 0,
            timeOfLastDragEvent: 0,
            touchStartX: 0,
            prevTouchX: 0,
            beingTouched: false,
            isOpen: false,
            touchStartY: 0,
        });
    });
});
