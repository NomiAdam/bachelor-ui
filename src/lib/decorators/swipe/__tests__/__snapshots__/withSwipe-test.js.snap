// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withSwipe should match the snapshot 1`] = `
<styled.div
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
>
  <[object Object]
    handleClick={[Function]}
    left={0}
  />
</styled.div>
`;
